cluster/kubernetes/sync.go

```diff
@@ -2,36 +2,327 @@ package kubernetes
+				c.logger.Log("warning", "cluster resource with empty stack label; skipping", "resource", resourceID)
+				continue
+			case actual != expected:
+				c.logger.Log("warning", "resource to be synced has not been updated; skipping", "resource", resourceID)
+				continue
+			default:
+				// The stack and checksum are the same, indicating that it was applied earlier. Leave it alone.
+			}
+		}
+
+		if deleteErrs := c.applier.apply(logger, orphanedResources, nil); len(deleteErrs) > 0 {
+			errs = append(errs, deleteErrs...)
+		}
+	}
+
+	// If `nil`, errs is a cluster.SyncError(nil) rather than error(nil), so it cannot be returned directly.
+	if errs == nil {
+		return nil
+	}
+
+	// It is expected that Cluster.Sync is invoked with *all* resources.
+	// Otherwise it will override previously recorded sync errors.
+	c.setSyncErrors(errs)
+	return errs
+}
+
+// --- internals in support of Sync
+
+type kuberesource struct {
+	obj *unstructured.Unstructured
+}
+
+// ResourceID returns the ResourceID for this resource loaded from the
+// cluster.
+func (r *kuberesource) ResourceID() flux.ResourceID {
+	ns, kind, name := r.obj.GetNamespace(), r.obj.GetKind(), r.obj.GetName()
+	return flux.MakeResourceID(ns, kind, name)
+}
+
+// AssumedNamespaceResourceID returns a ResourceID which assumes the
+// namespace is fallbackNamespace if it is missing. Resources returned
+// from the cluster that are not namespaced (e.g., ClusterRoles) will
+// have an empty string in the namespace field. To be able to compare
+// them to resources we load from files, we must assume a namespace if
+// none is given.
+func (r *kuberesource) AssumedNamespaceResourceID(fallbackNamespace string) flux.ResourceID {
+	ns := r.obj.GetNamespace()
+	if ns == "" {
+		ns = fallbackNamespace
+	}
+	return flux.MakeResourceID(ns, r.obj.GetKind(), r.obj.GetName())
+}
+
+// Bytes returns a byte slice description
+func (r *kuberesource) Bytes() []byte {
+	return []byte(fmt.Sprintf(`
+apiVersion: %s
+kind: %s
+metadata:
+  namespace: %q
+  name: %q
+`, r.obj.GetAPIVersion(), r.obj.GetKind(), r.obj.GetNamespace(), r.obj.GetName()))
+}
+
+func (r *kuberesource) Policy() policy.Set {
+	return kresource.PolicyFromAnnotations(r.obj.GetAnnotations())
+}
+
+// GetChecksum returns the checksum recorded on the resource from
+// Kubernetes, or an empty string if it's not present.
+func (r *kuberesource) GetChecksum() string {
+	return r.obj.GetAnnotations()[checksumAnnotation]
+}
+
+// GetStack returns the stack recorded on the the resource from
+// Kubernetes, or an empty string if it's not present.
+func (r *kuberesource) GetStack() string {
+	return r.obj.GetLabels()[stackLabel]
+}
+
+func (c *Cluster) getResourcesBySelector(selector, assumedNamespace string) (map[string]*kuberesource, error) {
+	listOptions := meta_v1.ListOptions{}
+	if selector != "" {
+		listOptions.LabelSelector = selector
+	}
+
+	resources, err := c.client.coreClient.Discovery().ServerResources()
```

Comment:
AFAIU this is not cached, which is pretty expensive (~50 requests to the API server)

Reply:
We should try to find an approach which lets use a cached client (or minimize requests) also satisfying the needs of #1668